update optimistic transaction (#2824)
pingcap/docs: optimistic-transaction.md

# TiDB Optimistic Transaction Model

This document introduces the principles of TiDB's optimistic transaction model and related features.

> TiDB uses the optimistic transaction model by default. In TiDB's optimistic transaction model, the two-phase commit begins only after the transaction is committed.

Reviewer: The second sentence is not aligned with https://github.com/pingcap/docs-cn/pull/3171/files @ran-huang

Author: Updated.

> **Note:**
>

> > Starting from v3.0.8, the newly created TiDB clusters use the [pessimistic transaction model](/pessimistic-transaction.md) by default. However, this does not affect your clusters if you upgrading from v3.0.7 or earlier to v3.0.8 (and later). In other words, **only newly created clusters default to using the pessimistic transaction model**.

Reviewer: ```suggestion
> Starting from v3.0.8, the newly created TiDB cluster uses the [pessimistic transaction model](/pessimistic-transaction.md) by default. However, this does not affect your existing cluster if you upgrade it from v3.0.7 or earlier to v3.0.8 or later. In other words, **only newly created clusters default to using the pessimistic transaction model**.
```

## Principles of optimistic transactions

> To support distributed transactions, TiDB adopts two-phase commit (2PC) in the optimistic transaction model. The procedure is as follows:

Reviewer: ```suggestion
To support distributed transactions, TiDB adopts two-phase commit (2PC) in optimistic transactions. The procedure is as follows:
```

![2PC in TiDB](/media/2pc-in-tidb.png)

1. The client begins a transaction.

>     TiDB receives the timestamp (monotonically increasing in time and globally unique) from PD as the unique transaction ID of the current transaction, which is called `start_ts`. TiDB implements multi-version concurrency control, so `start_ts` also serves as the database snapshot version obtained by this transaction. The transaction can only read the data available for the `start_ts` version.

Reviewer: ```suggestion
    TiDB gets a timestamp (monotonically increasing in time and globally unique) from PD as the unique transaction ID of the current transaction, which is called `start_ts`. TiDB implements multi-version concurrency control, so `start_ts` also serves as the version of the database snapshot obtained by this transaction. This means that the transaction can only read the data from the database at `start_ts`.
```
PTAL at my suggestion and see if it's technically correct. @jackysp

Reviewer 2: It looks OK.

2. The client issues a read request.

    1. TiDB receives routing information (how data is distributed among TiKV nodes) from PD.
    2. TiDB receives the data of the `start_ts` version from TiKV.

3. The client issues a write request.

    TiDB checks whether the written data satisfies constraints (to ensure the data types are correct, the NOT NULL constraint is met, etc.). **Valid data is stored in the private memory of this transaction in TiDB**.

4. The client issues a commit request.

> 5. TiDB begins 2PC to ensure the atomicity of transactions and persist data in store.

Reviewer: ```suggestion
5. TiDB begins 2PC, and persist data in store while guaranteeing the atomicity of transactions.
```

However, TiDB transactions also have the following disadvantages:

* Transaction latency due to 2PC
* In need of a centralized timestamp allocation service
* OOM (out of memory) when extensive data is written in the memory

## Transaction retries

> In the optimistic transaction model, transactions might fail to be committed because of write–write conflict in heavy contention scenarios. TiDB uses optimistic concurrency control by default, whereas MySQL applies pessimistic concurrency control. This means that MySQL adds locks during the execution of a SQL statement, and its Repeatable Read isolation level allows for non-repeatable reads, so commits generally do not encounter exceptions. To lower the difficulty of adapting applications, TiDB provides a retry function that runs inside the database.

Reviewer: ```suggestion
In the optimistic transaction model, transactions might fail to be committed because of write–write conflict in heavy contention scenarios. TiDB uses optimistic concurrency control by default, whereas MySQL applies pessimistic concurrency control. This means that MySQL adds locks during SQL execution, and its Repeatable Read isolation level allows for non-repeatable reads, so commits generally do not encounter exceptions. To lower the difficulty of adapting applications, TiDB provides an internal retry mechanism.
```

### Automatic retry

> If write-write conflict occurs during a transaction commit, TiDB retries the write operations automatically. You can enable or disable this default function by setting `tidb_disable_txn_auto_retry` to `off` and set the retry limit by configuring `tidb_retry_limit`:

Reviewer: ```suggestion
If a write-write conflict occurs during the transaction commit, TiDB automatically retries the SQL statement that includes write operations. You can enable the automatic retry by setting `tidb_disable_txn_auto_retry` to `off` and set the retry limit by configuring `tidb_retry_limit`:
```

## Conflict detection

> Because TiDB is a distributed database, the conflict detection in the memory is performed in the TiKV layer, mainly in the prewrite phase. TiDB instances are stateless and unaware of each other, which means they cannot confirm whether there are conflicts or not. Therefore, the conflict detection is performed in the TiKV layer.

Reviewer: ```suggestion
As a distributed database, TiDB performs in-memory conflict detection in the TiKV layer, mainly in the prewrite phase. TiDB instances are stateless and unaware of each other, which means they cannot know whether their writes result in conflicts across the cluster. Therefore, conflict detection is performed in the TiKV layer.
```
